### DISCUSSION TITLE
How do I decide when to render react components on the server and when to render them on the client?

### DISCUSSION THREAD

**Discussion by amite** [vercel/next.js#18041] [Help] [answered]
I am new to server side rendering (using next.js). I want to understand what should I take into consideration while deciding what kind of components to render on the server and what kind client side. Any advice on how to go about finding out about this?

Also related: does the decision to statically render or render client side only take place at the top page level? Or can I do it at the component level as well? 

For example, can I have a dashboard page  which has the shell (sidebar and navbar) statically rendered and the components in the content area rendered client side via api?

**Comment by alanqchen** [comment, marked as answer]
For an introduction to the different options available to you, [this blog post (and the video)](https://www.jackherrington.com/csr-ssr-and-ssg-on-nextjs), I find to be helpful. Specifically for fetching from an API, [I participated in a discussion about this here](https://github.com/vercel/next.js/discussions/16883#discussioncomment-63404).

And yes, it's entirely possible to do what you have suggested, and I have done it myself for an [admin dashboard](https://github.com/alanqchen/Bear-Blog-Engine/blob/master/frontend/pages/auth/portal/dashboard/index.js).

Here's a picture of the initial page returned by the server, which is unpopulated:
<img src="https://user-images.githubusercontent.com/18543142/96557787-bff09780-1288-11eb-87f4-d412af7b5e89.png" width="20%" />

Vercel also does the same, with some cool skeleton placeholders while the data is being fetched:
<img src="https://user-images.githubusercontent.com/18543142/96558667-0266a400-128a-11eb-90ea-c3fb47e12610.png" width="40%" />

But _very_ generally, (and this is my opinion) for pages where the data needs to be up-to-date and/or dynamic, such as dashboards, I would render the skeleton layout a static page and go with client-side fetching. For pages that need good SEO and a low TTFB, use incremental static generation. But if the data does not change frequently or it's not important for the user to see the most up-to-date version, and you’re okay with rebuilding the site to update content, use `getStaticProps`.

As a side-note, I would also take a look at [SWR](https://swr.vercel.app/) when fetching data client-side as it provides some nice functions you may find useful.

I hope this helps!

**Reply by amite** [reply]
@alanqchen In your admin dashboard, [index.js](https://github.com/alanqchen/Bear-Blog-Engine/blob/master/frontend/pages/auth/portal/dashboard/index.js) page I don't see any data being passed at the top of the component tree. But I can see those calls to `getServerSideProps` on slug pages.

What does that mean? How does this achieve a static  skeleton layout? I am a bit confused. Can you clarify a bit more from examples from your project?

**Reply by alanqchen** [reply]
In my dashboard index page, I have a `PostsList` component, which contains a [call to dispatch a redux action](https://github.com/alanqchen/Bear-Blog-Engine/blob/master/frontend/components/PostsTable/postsTable.js#L72) which then will fetch the data from the API. Note that this is done on the client-side. So the page achieves a static skeleton layout since the page doesn't fetch data server-side using `getStaticProps`, `getServerSideProps`, etc. 

Regarding the slug page, that's an artifact of my old way of fetching the post from the server instead of the client, so I should move to get the query object from the server to the client as well and remove the `getServerSideProps` which right now just causes unnecessary server-side regenerations.

**Reply by amite** [reply]
Thank you. I have enough to give this a try by creating my own demo now
